File: 2.linked-list/single-linked-list/intersection-node/intersection_node.rb
def intersection_node(headA, headB)
  if headA.nil? || headB.nil?
    return nil
  end
  # 2 pointers for 2 list
  pa = headA
  pb = headB

  while pa != pb
    pa = pa.nil? ? headB : pa.next
    pb = pb.nil? ? headA : pb.next
  end
  return pa
end

File: 2.linked-list/single-linked-list/intersection-node/intersectionNode.js
let intersectionNode = function (headA, headB) {
  if (!headA || !headB) {
    return null;
  }
  let pa = headA;
  let pb = headB; // 2 pointers
  while (pa !== pb) {
    pa = pa ? pa.next : headB;
    pb = pb ? pb.next : headA;
  }
  return pa;
};
